### PR TITLE
SONARJAVA-4848 S6916 Use guard instead of a single if else in pattern match

### PR DESCRIPTION
**Pull request by leonardo-pilastri-sonarsource**
Please ensure your pull request adheres to the following guidelines: 

- [ ] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [ ] Unit tests are passing and you provided a unit test for your fix
- [ ] ITs should pass : To run ITs locally, checkout the [README](https://github.com/SonarSource/sonar-java/blob/master/README.md) of the project.
- [ ] If there is a [Jira](http://jira.sonarsource.com/browse/SONARJAVA) ticket available, please make your commits and pull request start with the ticket number (SONARJAVA-XXXX)

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/SingleIfInsteadOfPatternMatchGuardCheck.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CASE_GROUP);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    var caseGroup = (CaseGroupTree) tree;
+    var caseLabel = caseGroup.labels().get(0);
```
Comment (irina-batinic-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/SingleIfInsteadOfPatternMatchGuardCheck.java:57]: WDYT if we move this check down?
I think we might optimize a bit..

File: sonar-java-plugin/src/main/resources/org/sonar/l10n/java/rules/java/S6916.html [hunk 2]
```diff
@@ -0,0 +1,66 @@
+<h2>Why is this an issue?</h2>
+<p>In Java 21 switch statements and expressions are extended to work on any type, and case labels now also allow patterns instead of just
```
Comment (irina-batinic-sonarsource) [inline sonar-java-plugin/src/main/resources/org/sonar/l10n/java/rules/java/S6916.html:2]: I would rewrite it a bit:

> Java 21 has introduced enhancements to switch statements and expressions, allowing them to operate on any type, not just specific ones, as in previous versions. Furthermore, case labels have been upgraded to support patterns, providing an alternative to the previous restriction of only accepting constants.

**Review by irina-batinic-sonarsource — CHANGES_REQUESTED** [review]
Overall, the implementation looks good. Nice job!
I left two comments, feel free to have a look.

File: java-checks/src/test/java/org/sonar/java/checks/SingleIfInsteadOfPatternMatchGuardCheckTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import org.junit.jupiter.api.Test;
+import org.sonar.java.checks.verifier.TestUtils;
+import org.sonar.java.checks.verifier.internal.InternalCheckVerifier;
+
+class SingleIfInsteadOfPatternMatchGuardCheckTest {
```
[the remark this answers is not in the document]
Comment (leonardo-pilastri-sonarsource) [reply java-checks/src/test/java/org/sonar/java/checks/SingleIfInsteadOfPatternMatchGuardCheckTest.java:26]: Before java 21 the code would not compile as you would not have those kind of guarded type patterns

File: sonar-java-plugin/src/main/resources/org/sonar/l10n/java/rules/java/S6916.html (hunk 2 above)
Comment (irina-batinic-sonarsource) [reply sonar-java-plugin/src/main/resources/org/sonar/l10n/java/rules/java/S6916.html:2]: PR: https://github.com/SonarSource/rspec/pull/3692

**Pull request merged**